### ISSUE TITLE
Amiga large drive compatibility issues

### ISSUE THREAD

**Issue by thepillock** [ZuluSCSI/ZuluSCSI-firmware#6]
Tried to install a ZuluSCSI to my Amiga(s).
Small images work fine, 1GB and 4GB do not present any problems. They're using the default 512 block size.
Trying to use an 8GB image, it appears to work (I can partition the "drive" in HDToolbox and then format it) but if I write almost any amount of data to it, even just 100MB or so, when I reboot the drive has lost its information. It feels like the RDB is being corrupted once data is written so far into the disk, the drive name has been changed to something like DH1\o\o\o\o\o\" and it's unreadable.

I've tried changing block size, and using the appropriate filename such as HD20_4096.img

- Tried on two different ZuluSCSI devices, both with latest firmware (although issue existed on original firmware they shipped with)
- Tried on A4000 using CyberStorm SCSI controller, and A3000 using built-in scsi.device
- Both are running kickstart 3.2 which supports drives larger than 4GB
- The 3000 was running a SCSI2SD v5.2 and a single 16GB drive OK so should be no issues with the OS

Log file: 
[zululog.txt](https://github.com/ZuluSCSI/ZuluSCSI-firmware/files/8695435/zululog.txt)

**Comment by thepillock** [comment]
To provide a few more results of testing:

- I see block sizes of 4096 are not supported, so I've gone back to 512 for everything. Still not handling partitions over 4GB
- I took my SCSI2SD card from my A3000 (a 16GB card divided into partitions of 1GB, 3GB, 5GB and 6GB). Made an image of it, put it on an SD card. Using ZuluSCSI and that image, on exactly the same machine, the 1GB and 3GB partitions are available but 5GB and 6GB are corrupted - and that's using the same Kickstart ROMS which include the SCSI device drivers

**Comment by aperezbios** [comment]
Thanks for the report. We'll look in to it, and take action if we can reproduce. I see you're using exFAT. As a reminder for others, any volumes larger than 4GB must reside on an exFAT-formatted filesystem, since the maximum supported file size for FAT32 is 4GB.

**Comment by PetteriAimonen** [comment]
@thepillock The image taken from SCSI2SD being corrupted is useful datapoint. It could be useful to have a debug log from when ZuluSCSI tries to read that partition.

**Comment by PetteriAimonen** [comment]
@thepillock Actually there was a simple integer overflow issue that caused writes and reads to wrap over to the first 4 GB.

There is a new build for testing here, it also adds support for 4096 and 8192 sector sizes:
https://github.com/ZuluSCSI/ZuluSCSI-firmware/releases/tag/latest

**Comment by thepillock** [comment]
I've flashed that firmware and it's looking good on the A3000 - I've been able to write 200MB of data to an 8GB partition and it's still there after a reboot. A sector size of 4096 also seems to work, there's a notable performance increase. Previously, even just 20-30MB of data would corrupt the partition.

I'll stress-test it later by sticking the whole 6GB games library on it, and also try the same firmware on the ZuluSCSI that's in my A4000. 

Really appreciate the quick response!
